data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarder.java

```diff
@@ -24,20 +27,36 @@
 
 @DataPrepperPlugin(name = "peer_forwarder", type = PluginType.PROCESSOR)
 public class PeerForwarder extends AbstractPrepper<Record<ExportTraceServiceRequest>, Record<ExportTraceServiceRequest>> {
+    public static final String FORWARD_REQUEST_LATENCY_PREFIX = "forwardRequestLatency";
+    public static final String FORWARD_REQUEST_SUCCESS_PREFIX = "forwardRequestSuccess";
+    public static final String FORWARD_REQUEST_ERRORS_PREFIX = "forwardRequestErrors";
+    public static final String PEER_ENDPOINTS = "peerEndpoints";
+
     private static final Logger LOG = LoggerFactory.getLogger(PeerForwarder.class);
 
     private final HashRing hashRing;
     private final PeerClientPool peerClientPool;
     private final int maxNumSpansPerRequest;
 
+    private final Map<String, Timer> forwardRequestTimers;
+    private final Map<String, Counter> forwardedRequestCounters;
+    private final Map<String, Counter> forwardRequestErrorCounters;
+    private final PeerListProvider peerListProvider;
+
     public PeerForwarder(final PluginSetting pluginSetting,
+                         final PeerListProvider peerListProvider,
                          final PeerClientPool peerClientPool,
                          final HashRing hashRing,
                          final int maxNumSpansPerRequest) {
         super(pluginSetting);
         this.peerClientPool = peerClientPool;
         this.hashRing = hashRing;
         this.maxNumSpansPerRequest = maxNumSpansPerRequest;
+        this.peerListProvider = pluginMetrics.gauge(
+                PEER_ENDPOINTS, peerListProvider, provider -> (double) provider.getPeerList().size());
```

Comment:
Nice, do we know how this works? How/when is the gauge value updated?

Reply:
Could this be an integer instead of double?

Reply:
So the passed-in function will be invoked on the passed-in instance everytime Measurement.getValue is called, e.g.

```
@Test
    public void testUpdate() {
        Foo foo = new Foo();

        MetricsTestUtil.initMetrics();

        PLUGIN_METRICS.gauge("gauge", foo, obj -> obj.getSize());

        List<Measurement> res = MetricsTestUtil.getMeasurementList(new StringJoiner(MetricNames.DELIMITER)
                .add(PIPELINE_NAME).add(PLUGIN_NAME)
                .add("gauge").toString());
        Assert.assertEquals(0.0, res.get(0).getValue(), 0);

        foo.a.add("sth");
        Assert.assertEquals(1.0, res.get(0).getValue(), 0);
    }

    static class Foo {
        List<String> a = new ArrayList<>();

        public int getSize() {
            return a.size();
        }
    }
```
I was not able to mock PeerListProviderFactory::new in the context of PeerForwarder so there is no test case here to manifest that.

Reply:
> Nice, do we know how this works? How/when is the gauge value updated?

Gauge has a weak-reference and auto-updates.

Reply:
> Could this be an integer instead of double?

Actually the cast to double might not be necessary. But in test case `.getValue()` still returns double.